From 77d70c8fa5655b9d43e6fa54abfe0a52e07c41cd Mon Sep 17 00:00:00 2001
From: nyawox <git@dof.pet>
Date: Tue, 25 Mar 2025 01:45:29 +0900
Subject: [PATCH] Squashed commit of the following:
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

commit 4e18938daf7ce1da4233060749126feb7cff1085
Merge: ea31a5b1 8c68074f
Author: TornaxO7 <tornax@proton.me>
Date:   Sat Nov 25 15:21:53 2023 +0100

    Merge branch 'master' of github.com:helix-editor/helix into md/global-status-line

commit ea31a5b14ccd6525694a64831288c19c4aedfc0a
Author: Matouš Dzivjak <matousdzivjak@gmail.com>
Date:   Sat Oct 22 12:44:52 2022 +0200

    feat(view): global status line option

    Add option to render single glboal status line instead of rendering
    status line per view.

    Fixes: https://github.com/helix-editor/helix/issues/2254
---
 helix-term/src/ui/editor.rs | 57 ++++++++++++++++++++++++++++++-------
 helix-view/src/editor.rs    |  9 ++++++
 2 files changed, 55 insertions(+), 11 deletions(-)

diff --git a/helix-term/src/ui/editor.rs b/helix-term/src/ui/editor.rs
index c7d1a24a..228d10d0 100644
--- a/helix-term/src/ui/editor.rs
+++ b/helix-term/src/ui/editor.rs
@@ -24,8 +24,8 @@
 };
 use helix_view::{
     annotations::diagnostics::DiagnosticFilter,
-    document::{Mode, DEFAULT_LANGUAGE_NAME, SCRATCH_BUFFER_NAME},
-    editor::{CompleteAction, CursorShapeConfig, InlineBlameBehaviour},
+    document::{Mode, SavePoint, DEFAULT_LANGUAGE_NAME, SCRATCH_BUFFER_NAME},
+    editor::{CompleteAction, CursorShapeConfig, InlineBlameBehaviour, StatusLineRenderConfig},
     graphics::{Color, CursorKind, Modifier, Rect, Style},
     icons::ICONS,
     input::{KeyEvent, MouseButton, MouseEvent, MouseEventKind},
@@ -87,7 +87,10 @@ pub fn render_view(
         is_focused: bool,
     ) {
         let inner = view.inner_area(doc);
-        let area = view.area;
+        let area = match editor.config().statusline.render {
+            StatusLineRenderConfig::Single => view.area.clip_bottom(1),
+            StatusLineRenderConfig::PerView => view.area,
+        };
         let theme = &editor.theme;
         let config = editor.config();
 
@@ -241,7 +244,6 @@ pub fn render_view(
             for y in area.top()..area.bottom() {
                 surface[(x, y)]
                     .set_symbol(tui::symbols::line::VERTICAL)
-                    //.set_symbol(" ")
                     .set_style(border_style);
             }
         }
@@ -252,15 +254,31 @@ pub fn render_view(
             Self::render_diagnostics(doc, view, inner, surface, theme);
         }
 
-        let statusline_area = view
-            .area
-            .clip_top(view.area.height.saturating_sub(1))
-            .clip_bottom(1); // -1 from bottom to remove commandline
+        match editor.config().statusline.render {
+            StatusLineRenderConfig::PerView => {
+                let statusline_area = view
+                    .area
+                    .clip_top(view.area.height.saturating_sub(1))
+                    .clip_bottom(1); // -1 from bottom to remove commandline
 
-        let mut context =
-            statusline::RenderContext::new(editor, doc, view, is_focused, &self.spinners);
+                let mut context =
+                    statusline::RenderContext::new(editor, doc, view, is_focused, &self.spinners);
 
-        statusline::render(&mut context, statusline_area, surface);
+                statusline::render(&mut context, statusline_area, surface);
+            }
+            StatusLineRenderConfig::Single => {
+                // -1 for command line
+                if viewport.bottom() - 1 != view.area.bottom() {
+                    let y = area.bottom();
+                    let border_style = theme.get("ui.window");
+                    for x in area.left()..area.right() {
+                        surface[(x, y)]
+                            .set_symbol(tui::symbols::line::HORIZONTAL)
+                            .set_style(border_style);
+                    }
+                }
+            }
+        };
     }
 
     pub fn render_rulers(
@@ -1604,6 +1622,23 @@ fn render(&mut self, area: Rect, surface: &mut Surface, cx: &mut Context) {
             self.render_view(cx.editor, doc, view, area, surface, is_focused);
         }
 
+        if config.statusline.render == StatusLineRenderConfig::Single {
+            if let Some((view, is_focused)) =
+                cx.editor.tree.views().find(|&(_, is_focused)| is_focused)
+            {
+                let doc = cx.editor.document(view.doc).unwrap();
+                let mut context = statusline::RenderContext::new(
+                    cx.editor,
+                    doc,
+                    view,
+                    is_focused,
+                    &self.spinners,
+                );
+                let statusline_area = area.clip_top(area.height.saturating_sub(2)).clip_bottom(1); // -1 from bottom to remove commandline
+                statusline::render(&mut context, statusline_area, surface);
+            }
+        }
+
         if config.auto_info {
             if let Some(mut info) = cx.editor.autoinfo.take() {
                 info.render(area, surface, cx);
diff --git a/helix-view/src/editor.rs b/helix-view/src/editor.rs
index ad152870..9951c949 100644
--- a/helix-view/src/editor.rs
+++ b/helix-view/src/editor.rs
@@ -537,6 +537,7 @@ pub struct StatusLineConfig {
     pub right: Vec<StatusLineElement>,
     pub separator: String,
     pub mode: ModeConfig,
+    pub render: StatusLineRenderConfig,
 }
 
 impl Default for StatusLineConfig {
@@ -561,6 +562,7 @@ fn default() -> Self {
             ],
             separator: String::from("│"),
             mode: ModeConfig::default(),
+            render: StatusLineRenderConfig::PerView,
         }
     }
 }
@@ -583,6 +585,13 @@ fn default() -> Self {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum StatusLineRenderConfig {
+    Single,
+    PerView,
+}
+
 #[derive(Debug, Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
 #[serde(rename_all = "kebab-case")]
 pub enum StatusLineElement {
-- 
2.48.1

